accelerator/migrations/0019_add_deferred_user_role.py

```
@@ -0,0 +1,32 @@
+# Generated by Django 2.2.10 on 2020-04-09 21:24
+
+from django.db import migrations
+from django.db.utils import IntegrityError
+
+
+def add_deferred_user_role(apps, schema_editor):
+    DEFERRED_MENTOR = 'Deferred Mentor'
+    UserRole = apps.get_model('accelerator', 'UserRole')
+    Program = apps.get_model('accelerator', 'Program')
+    ProgramRole = apps.get_model('accelerator', 'ProgramRole')
+    user_role = UserRole.objects.create(
+        name=DEFERRED_MENTOR, sort_order='17')
+    for program in Program.objects.all():
+        try:
+            ProgramRole.objects.get_or_create(
+                program=program,
+                user_role=user_role)
+        except IntegrityError:
```

I'm curious about the IntegrityError. Did you have any idea of what was causing this to be raised?

---

When I ran `make migrate` I was getting an `django.db.utils.IntegrityError: (1062, "Duplicate entry '' for key 'name'")` and I realized that when a program role is created but the name is empty so that is why it is triggered.

---

Ah, then we need to deal with that. 

On clean-db, before migration:
```
In [3]: mc.ProgramRole.objects.count()
Out[3]: 1340
```
after migration: 
```
In [4]: mc.ProgramRole.objects.count()
Out[4]: 1341
```


Only one was created, and it was:
```
In [6]: mc.ProgramRole.objects.latest("id")
Out[6]: <ProgramRole: >
```
hence the integrity errors - it was trying to create multiple Program Roles with the same (blank) name. With the `except: pass`, we just don't create the Program Roles at all, which is not what we want. 

I suggest composing a name, such as
```
    name="{} {} Deferred Mentor".format(program.start_date.year, program.program_family.name)
```

(ideally, we'd be able to use the Program.url_slug field, but staff have not abided by our standards, so we cannot get the year from that slug as we'd expect to)

---

This explains it much better. Thanks for the suggestion. Trying it out currently